### PR TITLE
fix(cli): only count errors for max diagnostics

### PR DESCRIPTION
**Pull request by yossydev**
## Summary

Part of: 
- #1083 

## Test Plan

- Should I write a test?👀

### CODE REVIEW

**Comment by ematipico** [comment]
If you can, that would be great.

We need to test a situation where we have more warnings than errors, and make sure that all the errors are written.

**Comment by yossydev** [comment]
> If you can, that would be great.
> 
> We need to test a situation where we have more warnings than errors, and make sure that all the errors are written.

Thank you!!
I will try to write a test that creates a situation where there are more warnings than errors, and ensures that the errors are logged!

File: crates/biome_cli/src/execute/traverse.rs [hunk 1]
```diff
@@ -466,9 +473,23 @@ fn process_messages(options: ProcessMessagesOptions) {
                         let severity = diag.severity();
                         if severity == Severity::Error {
                             *errors += 1;
+                            let should_print = printed_error_diagnostics < max_diagnostics;
+                            if should_print {
```
Comment (unvalley) [inline crates/biome_cli/src/execute/traverse.rs:477]: Shadowing could be reduced.
```suggestion
                            if printed_error_diagnostics < max_diagnostics {
```

File: crates/biome_cli/src/execute/traverse.rs [hunk 2]
```diff
@@ -466,9 +473,23 @@ fn process_messages(options: ProcessMessagesOptions) {
...
+                                    Ordering::Relaxed,
+                                );
+                            } else {
+                                not_printed_diagnostics += 1;
+                            }
                         }
                         if severity == Severity::Warning {
                             *warnings += 1;
+                            let should_print = printed_warning_diagnostics < max_diagnostics;
+                            if should_print {
```
Comment (unvalley) [inline crates/biome_cli/src/execute/traverse.rs:490]:
```suggestion
                            if printed_warning_diagnostics < max_diagnostics {
```

**Comment by yossydev** [comment]
@ematipico 
> If you can, that would be great.
> 
> We need to test a situation where we have more warnings than errors, and make sure that all the errors are written.

Sorry, I tried to write a test, but I am still not very knowledgeable about Rust and it would take a long time, so if it is not mandatory, can I implement it separately from this PR...?? :eyes:

File: crates/biome_cli/src/execute/traverse.rs (hunk 1 above)
Comment (yossydev) [reply crates/biome_cli/src/execute/traverse.rs:477]: Thank you!
https://github.com/biomejs/biome/pull/1248/commits/a2d9c0f7ee28d993961b8f5daefb7eaa017b200a ← fixed it!

**Comment by ematipico** [comment]
Thank you @yossydev , that's fine and we appreciate your contribution.

You can implement the tests in a separate pr!

File: crates/biome_cli/src/execute/traverse.rs [hunk 3]
```diff
@@ -309,6 +309,8 @@ fn process_messages(options: ProcessMessagesOptions) {
 
     let mut paths: FxHashSet<String> = FxHashSet::default();
     let mut printed_diagnostics: u16 = 0;
+    let mut printed_error_diagnostics: u16 = 0;
+    let mut printed_warning_diagnostics: u16 = 0;
```
Comment (ematipico) [inline crates/biome_cli/src/execute/traverse.rs:313]: I see an issue here. Diagnostics aren't just warnings or errors. Diagnostics have **many** levels of severity:

https://github.com/biomejs/biome/blob/fb2bce9c4a14dd5c30a2dbae2ac5b01b6604b6db/crates/biome_diagnostics/src/diagnostic.rs#L121-L133

I think we should not keep count of warnings and errors.

I think the solution is simpler:

```rs
let shout_print = printed_diagnostics < max_diagnostics && diag.severity() >= *diagnostic_level;
if should_print {
	printed_diagnostics += 1
}
```

Essentially, before increasing `printed_diagnostics`, we should check if the diagnosing we received from the thread has a severity equals or higher than `diagnostic_level`, which is the level set by the user.

**Review by ematipico — CHANGES_REQUESTED** [review]

File: crates/biome_cli/src/execute/traverse.rs (hunk 3 above)
Comment (ematipico) [reply crates/biome_cli/src/execute/traverse.rs]: @yossydev yes, but you have to repeat this logic in all the places where you have `printed_diagnostics += 1`

Comment (yossydev) [reply crates/biome_cli/src/execute/traverse.rs]: @ematipico 
Thank you! I have added an additional fix with https://github.com/biomejs/biome/pull/1248/commits/195639ce668b8adfe4772af8048ded3c3818c6f5!

I just haven't been able to fix the other two...🙏🏻 

https://github.com/biomejs/biome/blob/28f2de482625e2ce650b407637981720d1298df9/crates/biome_cli/src/execute/traverse.rs#L347-L349

https://github.com/biomejs/biome/blob/28f2de482625e2ce650b407637981720d1298df9/crates/biome_cli/src/execute/traverse.rs#L505

Do I need to edit the `AppryError` and `Diff` structures to use `.severity()`?👀

Comment (ematipico) [reply crates/biome_cli/src/execute/traverse.rs]: Yes @yossydev  :)

**All** places where we increase the printed diagnostics

Comment (yossydev) [reply crates/biome_cli/src/execute/traverse.rs]: Thanks for your comment!

I did the rest of the modifications at https://github.com/biomejs/biome/pull/1248/commits/0869b03d92b298cd8b0ad0cefa120aeaf731d8c6. Is this the right place to do it?👀

Comment (yossydev) [reply crates/biome_cli/src/execute/traverse.rs]: @ematipico 
Thanks for your comment!

I did the rest of the modifications at https://github.com/biomejs/biome/pull/1248/commits/0869b03d92b298cd8b0ad0cefa120aeaf731d8c6. Is this the right place to do it?👀

**Review by ematipico — APPROVED** [review]
I will take over this PR, there are some changes that I have to do. Thank you for your contribution!

**Pull request merged**